### PR TITLE
db: add deletable value merger

### PR DESCRIPTION
**Pull request by damnever**
DeletableValueMerger allows the caller to indicate if the key is deletable.

Related issue #1219

### CODE REVIEW

File: iterator.go [hunk 1]
```diff
@@ -1235,8 +1257,10 @@ func (stats *IteratorStats) String() string {
 func (stats *IteratorStats) SafeFormat(s redact.SafePrinter, verb rune) {
 	for i := range stats.ForwardStepCount {
 		switch IteratorStatsKind(i) {
-		case InterfaceCall: s.SafeString("(interface (dir, seek, step): ")
-		case InternalIterCall: s.SafeString(", (internal (dir, seek, step): ")
+		case InterfaceCall:
```
Comment (damnever) [inline iterator.go:1260]: Maybe the `goimports` did this..

File: compaction_iter.go [hunk 2]
```diff
@@ -325,7 +326,11 @@ func (i *compactionIter) Next() (*InternalKey, []byte) {
 				// includesBase is true whenever we've transformed the MERGE record
 				// into a SET.
 				includesBase := i.key.Kind() == InternalKeyKindSet
```
Comment (damnever) [inline compaction_iter.go:328]: The `includesBase` seems always to be false if I only use the Merge operation.

**Comment by damnever** [comment]
@petermattis PTAL. Test cases have been added

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1230)* status: 0 of 8 files reviewed, 2 unresolved discussions (waiting on @damnever)

___
*[compaction_iter.go, line 328 at r2](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MhTWQimOzVKRyMkSVmv-r2-328:-MiWryG72eqe77KmuJYr:b42rs9a) ([raw file](https://github.com/cockroachdb/pebble/blob/810e92485a01b9f7c5bc5c22dff27199d26ee8ac/compaction_iter.go#L328)):*
> Thinking about this more, could you give a concrete example of the semantics of your merge operation such that you'd want to transform a merge into a delete. Perhaps the includesBase part isn't necessary.

@damnever Thanks for the test addition. I'm hoping you can still answer this question for a concrete example of a merge operator where you'd like to be able to transform a merge into a delete. Is the "deletable sum value merger" what you actually want this for, or is there some other merger? This is mostly for my own mental model of the utility of this new functionality.

PS It will be a few days before I can give this a proper review.

**Comment by damnever** [comment]
@petermattis There are two cases:

1. Inverted index: each key stores multiple uint64 values(encoded with [Roaring Bitmap](https://roaringbitmap.org/)), the index can be deleted, it **may** never be reused if the value becomes empty.
2. Assume there is a key `A` stores an uint64 value:
    1. The following case shows the key has been deleted for good (the merge operation isn't necessary here).
       a. `Merge(A, opSet:1) -> Merge(A, opDel:1)`
    2. But there is another case, the key may be updated:
        a. `Merge(A, opSet:1) -> Merge(A, opSet:2) -> Merge(A, opDel:1)`
        b. `Merge(A, opSet:1) -> Merge(A, opDel:1) -> Merge(A, opSet:2)`
        In this case, the number is guaranteed to be monotonically increased (also we can assume `opSet` is always executed in one thread, no overflows, and no `opDel` happens before corresponding `opSet`), if the `includeBase` is true, we can simply store the bigger value or nothing since bigger value wins, otherwise, we need to store the full version of values.

Our scenario is more writing than reading, a lock-protected CAS is too expensive.

**Review by petermattis — COMMENTED** [review]
Can you give a bit more color for the inverted index use case? Is the idea that you'd be adding "document IDs" or something like that to a single key, but also want the ability to indicate that they should be deleted. So you'd have `add(doc-id)` and `del(doc-id)` merge operations.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1230#-:-MiqdLpk-uSFfeNR1O6S:bdq24m0)* status: 0 of 8 files reviewed, 11 unresolved discussions (waiting on @damnever)

___
*[compaction_iter.go, line 334 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-Miqd0wM6ADcgN82unVZ:-Miqd0wM6ADcgN82unV_:bo2ngin) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/compaction_iter.go#L334)):*
> ```Go
> 			if i.err == nil {
> 				if needDelete {
> 					return i.Next()
> ```

Can the recursive call here be replaced with a `continue`? Is there any other state that needs to be reset?
___
*[iterator.go, line 1260 at r1](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MhTWQimOzVKRyMkSVmx-r1-1260:-MiqbHkuD2g4FA74qRqf:b-k9m3f2) ([raw file](https://github.com/cockroachdb/pebble/blob/c28663b3a352aa4bf5be285f12360caf91b57461/iterator.go#L1260)):*
<details><summary><i>Previously, damnever (Xiaochao Dong) wrote…</i></summary><blockquote>

Maybe the `goimports` did this..
</blockquote></details>

Possibly. We format Pebble code using `crlfmt` from the CRDB repo. 
___
*[iterator.go, line 258 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MiqcO_wBpg8IqGKVNHv:-MiqcO_wBpg8IqGKVNHw:b-82pggm) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/iterator.go#L258)):*
> ```Go
> 		case InternalKeyKindMerge:
> 			var valueMerger ValueMerger
> 			valueMerger, i.err = i.merge(key.UserKey, i.iterValue)
> ```

Huh? Was that a bug before?
___
*[iterator.go, line 270 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-Miqcddr4Nc9RKpuarf9:-Miqcddr4Nc9RKpuarfA:b-uj1eg7) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/iterator.go#L270)):*
> ```Go
> 						i.nextUserKey()
> 					}
> 					i.findNextEntry(limit)
> ```

Recursively calling `findNextEntry` is odd. Is there a reason you can't `continue` here.
___
*[iterator.go, line 444 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MiqclDv2-MyacFxRjyB:-MiqclDv2-MyacFxRjyC:bdaib12) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/iterator.go#L444)):*
> ```Go
> 						i.value = nil
> 						i.iterValidityState = IterExhausted
> 						i.findPrevEntry(limit)
> ```

Ditto about the recursive all here. Can you `continue` instead.
___
*[iterator_test.go, line 412 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-Miqaq01AW5oAOzL59wK:-Miqaq01AW5oAOzL59wL:bheh50r) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/iterator_test.go#L412)):*
> ```Go
> }
> 
> type deletableSumValueMerger struct {
> ```

Mind add a comment here around what the semantics of this merger are. Something like: `deletableSumValueMerger computes the sum of its arguments, but transforms a zero sum into a non-existent entry`.
___
*[merger.go, line 28 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MiqaPnNCCpsOWmm7gu4:-MiqaPnNCCpsOWmm7gu5:bt5bu7l) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/merger.go#L28)):*
> ```Go
> var DefaultMerger = base.DefaultMerger
> 
> func finishValueMerger(valueMerger ValueMerger, includesBase bool) (value []byte, needDelete bool, closer io.Closer, err error) {
> ```

Nit: can you reformat this as:

```
func finishValueMerger(
	valueMerger ValueMerger, includesBase bool,
) (value []byte, needDelete bool, closer io.Closer, err error) {
```

We try to keep line lengths less than 100 columns for better readability in side-by-side code review tools.
___
*[internal/base/merger.go, line 63 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MiqZ69A78IuMsctXngf:-MiqZ69B6-VNpgFjt7__:b1h12w) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/internal/base/merger.go#L63)):*
> ```Go
> }
> 
> // DeletableValueMerger is the same as ValueMerger, except it will hide the value if
> ```

`hide the value` isn't quite right. Some word smithing here:

```
DeletableValueMerger is an extension to ValueMerger which allows indicating that the
result of a merge operation is non-existent. Such non-existent entries will eventually
be deleted during compaction. Note that during compaction, non-existence of the result 
of a merge means that the merge operands will not result in any record being output.
This is not the same as transforming the merge operands into a deletion tombstone, as
older merge operands will still be visible during iteration. Deletion of the merge operands
in this way is akin to the way a SingleDelete+Set combine into non-existence while leaving
older records for the same key unaffected.
```
___
*[internal/base/merger.go, line 70 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MiqZSV61Xw4wZiNdJsn:-MiqZSV61Xw4wZiNdJso:b-gnirei) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/internal/base/merger.go#L70)):*
> ```Go
> 	// DeletableFinish enables caller to indicate if the key is deletable.
> 	// Same as Finish, caller must handle it carefully if the includesBase
> 	// is false since the compaction may only handle a subset of operands.
> ```

Do you mind if I word smith this a bit:

```
DeletableFinish enables a value merger to indicate that the result of a merge operation
is non-existent. See Finish for a description of includesBase.
```
___
*[internal/base/merger.go, line 71 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MiqZoNh6HN2To2tzUIf:-MiqZoNh6HN2To2tzUIg:b-zejf9c) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/internal/base/merger.go#L71)):*
> ```Go
> 	// Same as Finish, caller must handle it carefully if the includesBase
> 	// is false since the compaction may only handle a subset of operands.
> 	DeletableFinish(includesBase bool) ([]byte, bool, io.Closer, error)
> ```

With this number of return values, we should name them:

```
DeletableFinish(includesBase bool) (value []byte, delete bool, _ io.Closer, _ error)
```

**Comment by damnever** [comment]
Yes, the inverted index use case is just like what you described, I have defined multiple internal kinds to distinguish if it is a single add/del, or merged values/tombstones.

**Review by petermattis — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1230)* status: 0 of 8 files reviewed, 13 unresolved discussions (waiting on @damnever)

___
*[iterator.go, line 270 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MhTWQimOzVKRyMkSVmx-r4-270:-MjUiLwa2zbFGr00Zjhn:b-z9jcg3) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/iterator.go#L270)):*
<details><summary><i>Previously, damnever (Xiaochao Dong) wrote…</i></summary><blockquote>

The recursive call seems more proper because I can't find a better way to handle the `valueCloser`.
</blockquote></details>

I would expect `i.valueCloser` to be `nil` if `needDelete` is `true`. Can we enforced that in some manner?

**Comment by damnever** [comment]
@petermattis I have no idea since I usually write things like this:
```go
type valueMerger struct {}
func(m *valueMerger) Close() error { /*do some clean up and recycle work*/ }
func(m *valueMerger) DeletableFinish(...) ([]byte, bool, io.Closer, error) { return nil, true, m, nil }
```
and I'd expect that pebble will always do the clean up things for me.

**Review by petermattis — APPROVED** [review]
@damnever Apologies for the delay in reviewing. This PR slipped off my radar. Your new code looks good to go. Only one small question regarding the bug fix in `iterator.go`. Also, we typically structure a PR like this as a single commit. Can you squash all of your commits into one with a single commit message that describes the new functionality?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/pebble/1230#-:-MoctsS08douOJiQiZAA:b-zii7lk)* status: 0 of 8 files reviewed, 5 unresolved discussions (waiting on @damnever and @petermattis)

___
*[iterator.go, line 258 at r4](https://reviewable.io/reviews/cockroachdb/pebble/1230#-MhTWQimOzVKRyMkSVmx-r4-258:-MoctMCsC-eyk_OmFFhg:b4r1o5d) ([raw file](https://github.com/cockroachdb/pebble/blob/03f9b36abd96cda62f7fe827db322636ac9822d4/iterator.go#L258)):*
<details><summary><i>Previously, damnever (Xiaochao Dong) wrote…</i></summary><blockquote>

`i.key` is empty here, so this is a fix.
</blockquote></details>

I agree this looks like a good fix, but I'm disturbed that none of the tests fail with this change. Is there something in this PR that tests this fix? Can another test case of the normal merger be added which demonstrates the fix?

**Comment by damnever** [comment]
> I agree this looks like a good fix, but I'm disturbed that none of the tests fail with this change. Is there something in this PR that tests this fix? Can another test case of the normal merger be added which demonstrates the fix?

@petermattis As far as I can see, no mergers are using the key,  we may check the key in the test cases.

**Comment by petermattis** [comment]
Thanks for the contribution, @damnever.

**Pull request merged**